### PR TITLE
Bagatur/oai assistant

### PR DESCRIPTION
**Pull request by baskaryan**

### CODE REVIEW

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        raise AttributeError(
+            "Please make sure you are using a v1.1-compatible version of openai. You "
+            'can install with `pip install "openai>=1.1"`.'
+        ) from e
+
+
+class OpenAIAssistantRunnable(RunnableSerializable[Dict, Any]):
+    """Run an OpenAI Assistant."""
+
+    client: openai.OpenAI = Field(default_factory=_get_openai_client)
```
[the remark this answers is not in the document]
Comment (eyurtsev) [reply libs/experimental/langchain_experimental/openai_assistant/base.py:59]: Will we add async client as well?

[the remark this answers is not in the document]
Comment (baskaryan) [reply libs/experimental/langchain_experimental/openai_assistant/base.py:59]: ill be honest i still don't understand forward refs. will it? should i just type as Any?

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            name=name,
+            instructions=instructions,
+            tools=openai_tools,
+            model=model,
+        )
+        return cls(assistant_id=assistant.id, **kwargs)
+
+    @root_validator()
+    def validate_environment(cls, values: Dict) -> Dict:
+        if not values["client"]:
```
Comment (eyurtsev) [inline libs/experimental/langchain_experimental/openai_assistant/base.py:110]: Is this still neeeded if there's a default factory?

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+
+
+class OpenAIAssistantFinish(AgentFinish):
+    run_id: str
+    thread_id: str
+
+
+class OpenAIAssistantAction(AgentAction):
+    tool_call_id: str
+    run_id: str
```
Comment (eyurtsev) [inline libs/experimental/langchain_experimental/openai_assistant/base.py:28]: Could you document why run_id is here? (I assume just for debugging purposes -- i.e., if someone needs to access it to see what's going on?)

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 4]
```diff
@@ -0,0 +1,228 @@
...
+        if not values["client"]:
+            try:
+                import openai
+            except ImportError as e:
+                raise ImportError() from e
+
+            values["client"] = openai.OpenAI()
+        return values
+
+    def invoke(self, input: dict, config: Optional[RunnableConfig] = None) -> Any:
```
Comment (eyurtsev) [inline libs/experimental/langchain_experimental/openai_assistant/base.py]: would be easier to understand if the input: dict is typed

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 5]
```diff
@@ -0,0 +1,250 @@
...
+            if not self.as_agent:
+                return new_messages
+            answer: Any = [
+                msg_content for msg in new_messages for msg_content in msg.content
+            ]
+            if all(
+                isinstance(content, openai.types.beta.threads.MessageContentText)
+                for content in answer
+            ):
+                answer = "".join(content.text.value for content in answer)
```
Comment (bracesproul) [inline libs/experimental/langchain_experimental/openai_assistant/base.py:212]: Wouldn't it make more sense to join with newlines or at least spaces?

File: libs/experimental/langchain_experimental/openai_assistant/base.py (hunk 4 above)
Comment (baskaryan) [reply libs/experimental/langchain_experimental/openai_assistant/base.py]: typed as what? Dict[str, Any]?

File: libs/experimental/langchain_experimental/openai_assistant/base.py (hunk 3 above)
Comment (baskaryan) [reply libs/experimental/langchain_experimental/openai_assistant/base.py:28]: no needed for agent with custom tools, will document

File: libs/experimental/langchain_experimental/openai_assistant/base.py (hunk 2 above)
Comment (baskaryan) [reply libs/experimental/langchain_experimental/openai_assistant/base.py:186]: nope

File: libs/experimental/langchain_experimental/openai_assistant/base.py [hunk 6]
```diff
@@ -0,0 +1,341 @@
...
+            try:
+                import openai
+            except ImportError as e:
+                raise ImportError() from e
+
+            values["client"] = openai.OpenAI()
+        return values
+
+    def invoke(
+        self, input: dict, config: Optional[RunnableConfig] = None
```
Comment (eyurtsev) [inline libs/experimental/langchain_experimental/openai_assistant/base.py:189]: TypedDict would be nice here -- also useful for type inference

**Review by eyurtsev — APPROVED** [review]

**Pull request merged**